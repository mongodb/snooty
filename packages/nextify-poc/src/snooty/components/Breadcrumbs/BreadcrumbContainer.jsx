import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { reportAnalytics } from '../../utils/report-analytics';
import { theme } from '../../theme/docsTheme';
import { getFullBreadcrumbPath } from '../../utils/get-complete-breadcrumb-data';
import { useSiteMetadata } from '../../hooks/use-site-metadata';
import IndividualBreadcrumb from './IndividualBreadcrumb';
import CollapsedBreadcrumbs from './CollapsedBreadcrumbs';

const StyledSlash = styled('span')`
  cursor: default;
  padding-left: ${theme.size.small};
  padding-right: ${theme.size.small};
`;

const Flexbox = styled('div')`
  display: flex;
  align-items: center;
`;

const MIN_BREADCRUMBS = 3;
const initialMaxCrumbs = (breadcrumbs) => breadcrumbs.length + 1;

const BreadcrumbContainer = ({ breadcrumbs }) => {
  const [maxCrumbs, setMaxCrumbs] = React.useState(initialMaxCrumbs(breadcrumbs));
  const { siteUrl } = useSiteMetadata();

  React.useEffect(() => {
    const handleResize = () => {
      setMaxCrumbs(initialMaxCrumbs(breadcrumbs));
    };

    window.addEventListener('resize', handleResize);

    return () => window.removeEventListener('resize', handleResize);
  }, [breadcrumbs]);

  // Our breadcrumbs representation is an array of crumbObjectShape || (array of crumbObjectShape)
  // The latter indicates a collapsed series of breadcrumbs.
  const processedBreadcrumbs = React.useMemo(() => {
    const crumbsCopy = Array.from(breadcrumbs);
    if (crumbsCopy.length >= maxCrumbs && crumbsCopy.length > 2) {
      // A maximum of maxCrumbs breadcrumbs may be shown, so we collapse the first run of internal
      // crumbs into a single "â€¦" crumb
      const collapsedCrumbs = crumbsCopy.splice(1, breadcrumbs.length - maxCrumbs + 1, []);
      crumbsCopy[1] = collapsedCrumbs;
    }
    return crumbsCopy;
  }, [maxCrumbs, breadcrumbs]);

  const collapseBreadcrumbs = () => {
    const newMaxCrumbs = Math.max(maxCrumbs - 1, MIN_BREADCRUMBS);
    setMaxCrumbs(newMaxCrumbs);
  };

  return (
    <Flexbox>
      {processedBreadcrumbs.map((crumb, index) => {
        const isFirst = index === 0;
        return (
          <React.Fragment key={index}>
            {!isFirst && <StyledSlash> / </StyledSlash>}
            {Array.isArray(crumb) ? (
              <CollapsedBreadcrumbs crumbs={crumb}></CollapsedBreadcrumbs>
            ) : (
              <IndividualBreadcrumb
                key={crumb.title}
                crumb={crumb}
                setIsExcessivelyTruncated={collapseBreadcrumbs}
                onClick={() =>
                  reportAnalytics('BreadcrumbClick', {
                    breadcrumbClicked: getFullBreadcrumbPath(siteUrl, crumb.path, true),
                  })
                }
              ></IndividualBreadcrumb>
            )}
          </React.Fragment>
        );
      })}
    </Flexbox>
  );
};

const crumbObjectShape = {
  title: PropTypes.string.isRequired,
  path: PropTypes.string.isRequired,
};

BreadcrumbContainer.propTypes = {
  breadcrumbs: PropTypes.arrayOf(PropTypes.shape(crumbObjectShape)).isRequired,
};

export default BreadcrumbContainer;
