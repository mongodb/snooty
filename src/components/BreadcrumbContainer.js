import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { uiColors } from '@leafygreen-ui/palette';
import { css } from '@emotion/core';
import Link from './Link';
import { SNOOTY_STITCH_ID } from '../build-constants';
import { useSiteMetadata } from '../hooks/use-site-metadata';
import { theme } from '../theme/docsTheme';
import { formatText } from '../utils/format-text';
import { reportAnalytics } from '../utils/report-analytics';
import { fetchProjectParents } from '../utils/stitch';

const activeColor = css`
  color: ${uiColors.gray.dark3};
`;

const StyledNav = styled('nav')`
  font-size: ${theme.fontSize.small};
  min-height: ${theme.size.medium};

  * {
    color: ${uiColors.gray.dark1};
  }

  & > p {
    margin-top: 0;
  }
`;

const StyledArrow = styled('span')`
  cursor: default;

  :last-of-type {
    ${activeColor}
  }
`;

const StyledLink = styled(Link)`
  :last-of-type {
    ${activeColor}
  }

  :hover,
  :focus {
    text-decoration: none;

    :not(:last-of-type) {
      ${activeColor}
    }
  }
`;

const BreadcrumbContainer = ({ lastCrumb }) => {
  const { database, project } = useSiteMetadata();
  const [breadcrumbs, setBreadcrumbs] = useState([]);

  useEffect(() => {
    const fetchBreadcrumbData = async () => {
      let parentCrumbs = await fetchProjectParents(SNOOTY_STITCH_ID, database, project);
      const breadcrumbs = [
        { title: 'Docs Home', url: project === 'landing' ? '/' : 'https://docs.mongodb.com/' },
        ...parentCrumbs,
        lastCrumb,
      ];
      setBreadcrumbs(breadcrumbs);
    };
    fetchBreadcrumbData();
  }, [database, lastCrumb, project]);

  if (breadcrumbs.length === 0) {
    return null;
  }

  return (
    <StyledNav>
      <p>
        {breadcrumbs.map(({ title, url }, index) => {
          const isFirst = index === 0;
          return (
            <React.Fragment key={title}>
              {!isFirst && <StyledArrow> &#8594; </StyledArrow>}
              <StyledLink
                to={url}
                onClick={() => {
                  reportAnalytics('BreadcrumbClick', {
                    parentPaths: breadcrumbs,
                    breadcrumbClicked: url,
                  });
                }}
              >
                {formatText(title)}
              </StyledLink>
            </React.Fragment>
          );
        })}
      </p>
    </StyledNav>
  );
};

BreadcrumbContainer.propTypes = {
  lastCrumb: PropTypes.shape({
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
  }).isRequired,
};

export default BreadcrumbContainer;
