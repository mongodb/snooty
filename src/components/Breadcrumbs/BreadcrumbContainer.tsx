import React from 'react';
import styled from '@emotion/styled';
import { reportAnalytics } from '../../utils/report-analytics';
import { theme } from '../../theme/docsTheme';
import { getFullBreadcrumbPath } from '../../utils/get-complete-breadcrumb-data';
import { useSiteMetadata } from '../../hooks/use-site-metadata';
import IndividualBreadcrumb from './IndividualBreadcrumb';
import CollapsedBreadcrumbs from './CollapsedBreadcrumbs';

const StyledSlash = styled('span')`
  cursor: default;
  padding-left: ${theme.size.small};
  padding-right: ${theme.size.small};
`;

const Flexbox = styled('div')`
  display: flex;
  align-items: center;
`;

const MIN_BREADCRUMBS = 3;
const initialMaxCrumbs = (breadcrumbs: Array<BreadcrumbType>) => breadcrumbs.length + 1;

export type BreadcrumbType = {
  title: string;
  path: string;
};

const BreadcrumbContainer = ({ breadcrumbs }: { breadcrumbs: Array<BreadcrumbType> }) => {
  const [maxCrumbs, setMaxCrumbs] = React.useState(initialMaxCrumbs(breadcrumbs));
  const { siteUrl } = useSiteMetadata();

  React.useEffect(() => {
    const handleResize = () => {
      setMaxCrumbs(initialMaxCrumbs(breadcrumbs));
    };

    window.addEventListener('resize', handleResize);

    return () => window.removeEventListener('resize', handleResize);
  }, [breadcrumbs]);

  // Our breadcrumbs representation is an array of crumbObjectShape || (array of crumbObjectShape)
  // The latter indicates a collapsed series of breadcrumbs.
  const processedBreadcrumbs: (BreadcrumbType | BreadcrumbType[])[] = React.useMemo(() => {
    if (breadcrumbs.length >= maxCrumbs && breadcrumbs.length > 2) {
      // A maximum of maxCrumbs breadcrumbs may be shown, so we collapse the first run of internal
      // crumbs into a single "â€¦" crumb
      const crumbsCopy: (BreadcrumbType | BreadcrumbType[])[] = Array.from(breadcrumbs);
      const collapsedCrumbs = Array.from(breadcrumbs).splice(1, breadcrumbs.length - maxCrumbs + 1);
      const constructedCrumbs = Array.from(crumbsCopy).splice(1, breadcrumbs.length - maxCrumbs + 1, collapsedCrumbs);
      return constructedCrumbs;
    } else {
      return breadcrumbs;
    }
  }, [maxCrumbs, breadcrumbs]);

  const collapseBreadcrumbs = () => {
    const newMaxCrumbs = Math.max(maxCrumbs - 1, MIN_BREADCRUMBS);
    setMaxCrumbs(newMaxCrumbs);
  };

  return (
    <Flexbox>
      {processedBreadcrumbs.map((crumb, index) => {
        const isFirst = index === 0;
        return (
          <React.Fragment key={index}>
            {!isFirst && <StyledSlash> / </StyledSlash>}
            {Array.isArray(crumb) ? (
              <CollapsedBreadcrumbs crumbs={crumb}></CollapsedBreadcrumbs>
            ) : (
              <IndividualBreadcrumb
                key={crumb.title}
                crumb={crumb}
                setIsExcessivelyTruncated={collapseBreadcrumbs}
                onClick={() =>
                  reportAnalytics('BreadcrumbClick', {
                    breadcrumbClicked: getFullBreadcrumbPath(siteUrl, crumb.path, true),
                  })
                }
              ></IndividualBreadcrumb>
            )}
          </React.Fragment>
        );
      })}
    </Flexbox>
  );
};

export default BreadcrumbContainer;
