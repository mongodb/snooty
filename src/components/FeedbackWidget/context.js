import React from 'react';
import {
  createNewFeedback,
  updateFeedback,
  submitFeedback,
  abandonFeedback,
  useStitchUser,
  addAttachment,
} from './stitch';
import { getSegmentUserId } from './segment';
import { useScreenshot } from '../Screenshot';
import { getViewport } from '../../hooks/useViewport';
import * as R from 'ramda';

const FeedbackContext = React.createContext();

export function FeedbackProvider({ page, ...props }) {
  const [feedback, setFeedback] = React.useState(null);
  const [isSupportRequest, setIsSupportRequest] = React.useState(false);
  const [view, setView] = React.useState('waiting');
  const { screenshot, clearScreenshot } = useScreenshot();
  const user = useStitchUser();

  React.useEffect(() => {
    console.log('feedback', feedback);
    if (feedback) {
      console.log('feedback_id', feedback._id.toString());
    }
  }, [feedback]);

  async function initializeFeedback(initialView = 'rating') {
    const segment = getSegmentUserId();
    const newFeedback = await createNewFeedback({
      page: {
        title: page.title,
        slug: page.slug,
        url: page.url,
        docs_property: page.docs_property,
        docs_version: null,
      },
      user: {
        stitch_id: user.id,
        segment_id: segment.id,
        isAnonymous: segment.isAnonymous,
      },
      viewport: getViewport(),
    });
    setFeedback(newFeedback);
    setView(initialView);
    return newFeedback;
  }

  async function setRating(ratingValue) {
    let feedback_id;
    if (!feedback) {
      feedback_id = (await initializeFeedback('waiting'))._id;
    }
    if (feedback && feedback.rating) return;
    // Must be in range [1-5]
    if (typeof ratingValue !== 'number') {
      throw new Error('Rating value must be a number.');
    }
    if (ratingValue < 1 || ratingValue > 5) {
      throw new Error('Rating value must be between 1 and 5, inclusive.');
    }
    const updatedFeedback = await updateFeedback({
      feedback_id: feedback ? feedback._id : feedback_id,
      rating: ratingValue,
    });
    setFeedback(updatedFeedback);
    setView('qualifiers');
  }

  async function setQualifier(id, value) {
    if (!feedback) return;
    if (typeof id !== 'string') {
      throw new Error('id must be a string Qualifier ID.');
    }
    if (typeof value !== 'boolean') {
      throw new Error('value must be a boolean.');
    }
    const updatedQualifiers = R.adjust(
      feedback.qualifiers.findIndex(R.propEq('id', id)), // Find the qualifier by id
      q => ({ ...q, value }), // Update the value
      feedback.qualifiers // Adjust this array of qualifiers
    );
    const updatedFeedback = await updateFeedback({
      feedback_id: feedback._id,
      qualifiers: updatedQualifiers,
    });
    setFeedback(updatedFeedback);
  }

  function submitQualifiers() {
    if (!feedback) return;
    const selectedSupportQualifier = feedback.qualifiers.find(q => q.id === 'support' && q.value === true);
    setIsSupportRequest(Boolean(selectedSupportQualifier));
    setView('comment');
  }

  async function submitScreenshot({ dataUri, viewport }) {
    if (!feedback) return;
    const updatedFeedback = await addAttachment({
      feedback_id: feedback._id,
      attachment: { type: 'screenshot', dataUri, viewport },
    });
    setFeedback(updatedFeedback);
  }

  async function submitComment({ comment = '', email = '' }) {
    if (!feedback) return;
    await updateFeedback({
      feedback_id: feedback._id,
      comment,
      user: { email },
    });
    const submittedFeedback = await submitFeedback({ feedback_id: feedback._id });
    if (isSupportRequest) {
      setFeedback(submittedFeedback);
      setView('support');
    } else {
      setView('submitted');
      setFeedback(null);
    }
    screenshot && clearScreenshot();
    return submittedFeedback;
  }

  async function submitSupport() {
    if (!feedback) return;
    setView('submitted');
    setFeedback(null);
  }

  async function abandon() {
    if (feedback) {
      await abandonFeedback({ feedback_id: feedback._id });
    }
    screenshot && clearScreenshot();
    setView('waiting');
    setFeedback(null);
  }

  const value = {
    feedback,
    view,
    isSupportRequest,
    initializeFeedback,
    setRating,
    setQualifier,
    submitQualifiers,
    submitComment,
    submitScreenshot,
    submitSupport,
    abandon,
  };

  return <FeedbackContext.Provider value={value}>{props.children}</FeedbackContext.Provider>;
}

export function useFeedbackState() {
  const feedback = React.useContext(FeedbackContext);
  if (!feedback && feedback !== null) {
    throw new Error('You must nest useFeedbackState() inside of a FeedbackProvider.');
  }
  return feedback;
}
