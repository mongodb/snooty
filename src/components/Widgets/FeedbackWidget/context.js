import React from 'react';
import {
  createNewFeedback,
  updateFeedback,
  submitFeedback,
  abandonFeedback,
  useStitchUser,
  addAttachment,
} from './stitch';
import { getSegmentUserId } from '../../../utils/segment';
import { getViewport } from '../../../hooks/useViewport';

const FeedbackContext = React.createContext();

export function FeedbackProvider({ page, hideHeader, test = {}, ...props }) {
  const [feedback, setFeedback] = React.useState((test.feedback !== {} && test.feedback) || null);
  const [isSupportRequest, setIsSupportRequest] = React.useState(test.isSupportRequest || false);
  const [view, setView] = React.useState(test.view || 'waiting');
  const user = useStitchUser();

  // Create a new feedback document
  async function initializeFeedback(nextView = 'rating') {
    const segment = getSegmentUserId();
    const newFeedback = {
      page: {
        title: page.title,
        slug: page.slug,
        url: page.url,
        docs_property: page.docs_property,
        docs_version: null,
      },
      user: {
        stitch_id: user && user.id,
        segment_id: segment.id,
        isAnonymous: segment.isAnonymous,
      },
      viewport: getViewport(),
      ...test.feedback,
    };
    const { _id } = await createNewFeedback(newFeedback);

    setFeedback({ _id, ...newFeedback });
    setView(nextView);
    return { _id, ...newFeedback };
  }

  // Sets the user's star rating for the page
  async function setRating(ratingValue) {
    // Once a user has set a rating, they cannot change it unless they
    // abandon or submit the current feedback.
    if (feedback && feedback.rating) return;
    // Users on small screens start giving feedback by clicking a star
    // rating instead of the feedback tab. In this case, we need to
    // initialize a new feedback document before we set the rating.
    const feedback_id = feedback ? feedback._id : (await initializeFeedback('waiting'))._id;
    // The star rating must be in range [1-5]
    if (typeof ratingValue !== 'number') {
      throw new Error('Rating value must be a number.');
    }
    if (ratingValue < 1 || ratingValue > 5) {
      throw new Error('Rating value must be between 1 and 5, inclusive.');
    }
    // Update the feedback with the selected rating and then show the
    // user the relevant qualifier checkboxes. The qualifiers depend on
    // the rating so we need to await them from the server.
    const updatedFeedback = await updateFeedback({
      feedback_id: feedback ? feedback._id : feedback_id,
      rating: ratingValue,
    });
    setFeedback(updatedFeedback);
    setView('qualifiers');
  }

  // Sets the value of a single qualifier checkbox
  async function setQualifier(id, value) {
    if (!feedback) return;
    if (typeof id !== 'string') {
      throw new Error('id must be a string Qualifier ID.');
    }
    if (typeof value !== 'boolean') {
      throw new Error('value must be a boolean.');
    }
    const updatedFeedback = await updateFeedback({
      feedback_id: feedback._id,
      qualifiers: updateQualifier(feedback.qualifiers, id, value),
    });
    setFeedback(updatedFeedback);
  }

  // Once a user has selected qualifiers, show them the comment/email input boxes.
  function submitQualifiers() {
    if (!feedback) return;
    // The widget flow changes if the user selected the "need support" qualifier
    const selectedSupportQualifier = feedback.qualifiers.find((q) => q.id === 'support' && q.value === true);
    setIsSupportRequest(Boolean(selectedSupportQualifier));
    setView('comment');
  }

  // Upload a screenshot to S3 and attach it to the feedback
  async function submitScreenshot({ dataUri, viewport }) {
    if (!feedback) return;
    const updatedFeedback = await addAttachment({
      feedback_id: feedback._id,
      attachment: { type: 'screenshot', dataUri, viewport },
    });
    setFeedback(updatedFeedback);
  }

  // Submit the feedback and direct the user to the most appropriate "next steps" screen.
  async function submitComment({ comment = '', email = '' }) {
    if (!feedback) return;
    // Update the document with the user's comment and email (if provided)
    const updatedFeedback = await updateFeedback({
      feedback_id: feedback._id,
      comment,
      user: { email },
    });
    setFeedback(updatedFeedback);
  }

  async function submitAllFeedback() {
    // Submit the full feedback document
    const submittedFeedback = await submitFeedback({ feedback_id: feedback._id });
    setFeedback(submittedFeedback);
    // Route the user to their "next steps"
    if (isSupportRequest) {
      setView('support');
    } else {
      setView('submitted');
    }
  }

  // Show the user a thank you screen after they've seen support links
  async function submitSupport() {
    if (!feedback) return;
    setView('submitted');
  }

  // Stop giving feedback (if in progress) and reset the widget to the
  // initial state.
  async function abandon() {
    // Reset to the initial state
    setView('waiting');
    if (feedback) {
      // We hold on to abandoned feedback in the database, so wait until
      // we've marked the document as abandoned
      await abandonFeedback({ feedback_id: feedback._id });
      setFeedback(null);
    }
  }

  const value = {
    feedback,
    view,
    isSupportRequest,
    initializeFeedback,
    setRating,
    setQualifier,
    submitQualifiers,
    submitComment,
    submitScreenshot,
    submitSupport,
    submitAllFeedback,
    abandon,
    hideHeader,
  };

  return <FeedbackContext.Provider value={value}>{props.children}</FeedbackContext.Provider>;
}

function updateQualifier(qualifiers, id, value) {
  const index = qualifiers.findIndex((q) => q.id === id);
  return [...qualifiers.slice(0, index), { ...qualifiers[index], value }, ...qualifiers.slice(index + 1)];
}

export function useFeedbackState() {
  const feedback = React.useContext(FeedbackContext);
  if (!feedback && feedback !== null) {
    throw new Error('You must nest useFeedbackState() inside of a FeedbackProvider.');
  }
  return feedback;
}
