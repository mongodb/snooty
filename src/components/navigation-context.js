import React, { useEffect, useState } from 'react';
import { useSiteMetadata } from '../hooks/use-site-metadata';
import { isBrowser } from '../utils/is-browser';
import { fetchProjectParents } from '../utils/realm';

const NavigationContext = React.createContext({
  parents: [],
});

const NavigationProvider = ({ children }) => {
  const { database, project } = useSiteMetadata();
  const [parents, setParents] = useState([]);

  useEffect(() => {
    const fetchBreadcrumbData = async () => {
      try {
        const parents = await fetchProjectParents(database, project);
        setParents(parents);
      } catch (err) {
        console.error(err);
      }
    };
    if (isBrowser) {
      fetchBreadcrumbData();
    }
  }, [database, project]);

  return <NavigationContext.Provider value={{ parents }}>{children}</NavigationContext.Provider>;
};

export { NavigationContext, NavigationProvider };
