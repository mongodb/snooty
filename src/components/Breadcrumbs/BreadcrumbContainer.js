import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { uiColors } from '@leafygreen-ui/palette';
import { css } from '@emotion/react';
import Link from '../Link';
import { NavigationContext } from '../../context/navigation-context';
import { formatText } from '../../utils/format-text';
import { reportAnalytics } from '../../utils/report-analytics';

const activeColor = css`
  color: ${uiColors.gray.dark3};
`;

const StyledArrow = styled('span')`
  cursor: default;

  :last-of-type {
    ${activeColor}
  }
`;

const StyledLink = styled(Link)`
  :last-of-type {
    ${activeColor}
  }

  :hover,
  :focus {
    text-decoration: none;

    :not(:last-of-type) {
      ${activeColor}
    }
  }
`;

const BreadcrumbContainer = ({ homeCrumb, lastCrumb }) => {
  const { parents } = useContext(NavigationContext);
  const breadcrumbs = React.useMemo(() => [homeCrumb, ...parents, lastCrumb], [homeCrumb, parents, lastCrumb]);

  return (
    <>
      {breadcrumbs.map(({ title, url }, index) => {
        const isFirst = index === 0;
        return (
          <React.Fragment key={title}>
            {!isFirst && <StyledArrow> &#8594; </StyledArrow>}
            <StyledLink
              to={url}
              onClick={() => {
                reportAnalytics('BreadcrumbClick', {
                  parentPaths: breadcrumbs,
                  breadcrumbClicked: url,
                });
              }}
            >
              {formatText(title)}
            </StyledLink>
          </React.Fragment>
        );
      })}
    </>
  );
};

const crumbObjectShape = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};

BreadcrumbContainer.propTypes = {
  homeCrumb: PropTypes.shape(crumbObjectShape).isRequired,
  lastCrumb: PropTypes.shape(crumbObjectShape).isRequired,
};

export default BreadcrumbContainer;
