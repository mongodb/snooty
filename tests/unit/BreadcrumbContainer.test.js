import React from 'react';
import { render } from '@testing-library/react';
import BreadcrumbContainer from '../../src/components/Breadcrumbs/BreadcrumbContainer';
import { NavigationContext } from '../../src/context/navigation-context';

const mountBreadcrumbContainer = (homeCrumb, lastCrumb, parents) => {
  return render(
    <NavigationContext.Provider value={{ parents }}>
      <BreadcrumbContainer homeCrumb={homeCrumb} lastCrumb={lastCrumb} />
    </NavigationContext.Provider>
  );
};

describe('BreadcrumbContainer', () => {
  const mockHomeCrumb = {
    title: 'Docs Home',
    url: 'https://www.mongodb.com/docs/',
  };

  it('renders correctly with project parent', () => {
    const mockLastCrumb = {
      title: 'MongoDB Compass',
      url: 'documents/view',
    };

    const mockParents = [
      {
        title: 'View & Analyze',
        url: 'https://www.mongodb.com/docs/view-analyze',
      },
    ];

    const tree = mountBreadcrumbContainer(mockHomeCrumb, mockLastCrumb, mockParents);
    expect(tree.asFragment()).toMatchSnapshot();
  });

  it('renders correctly without project parent', () => {
    const mockLastCrumb = {
      title: 'View & Analyze Data',
      url: 'view-analyze',
    };

    const tree = mountBreadcrumbContainer(mockHomeCrumb, mockLastCrumb, []);
    expect(tree.asFragment()).toMatchSnapshot();
  });
});
