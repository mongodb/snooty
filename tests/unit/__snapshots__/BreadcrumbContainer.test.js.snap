// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbContainer renders correctly with project parent 1`] = `
<DocumentFragment>
  .emotion-0:last-of-type {
  color: #21313C;
}

.emotion-0:hover,
.emotion-0:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover:not(:last-of-type),
.emotion-0:focus:not(:last-of-type) {
  color: #21313C;
}

<a
    class="emotion-0 emotion-1"
    href="https://www.mongodb.com/docs/"
  >
    Docs Home
  </a>
  .emotion-0 {
  cursor: default;
}

.emotion-0:last-of-type {
  color: #21313C;
}

<span
    class="emotion-0 emotion-1"
  >
     → 
  </span>
  .emotion-0:last-of-type {
  color: #21313C;
}

.emotion-0:hover,
.emotion-0:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover:not(:last-of-type),
.emotion-0:focus:not(:last-of-type) {
  color: #21313C;
}

<a
    class="emotion-0 emotion-1"
    href="https://www.mongodb.com/docs/view-analyze"
  >
    View & Analyze
  </a>
  .emotion-0 {
  cursor: default;
}

.emotion-0:last-of-type {
  color: #21313C;
}

<span
    class="emotion-0 emotion-1"
  >
     → 
  </span>
  .emotion-0:last-of-type {
  color: #21313C;
}

.emotion-0:hover,
.emotion-0:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover:not(:last-of-type),
.emotion-0:focus:not(:last-of-type) {
  color: #21313C;
}

<a
    class="emotion-0 emotion-1"
    href="/documents/view/"
  >
    MongoDB Compass
  </a>
</DocumentFragment>
`;

exports[`BreadcrumbContainer renders correctly without project parent 1`] = `
<DocumentFragment>
  .emotion-0:last-of-type {
  color: #21313C;
}

.emotion-0:hover,
.emotion-0:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover:not(:last-of-type),
.emotion-0:focus:not(:last-of-type) {
  color: #21313C;
}

<a
    class="emotion-0 emotion-1"
    href="https://www.mongodb.com/docs/"
  >
    Docs Home
  </a>
  .emotion-0 {
  cursor: default;
}

.emotion-0:last-of-type {
  color: #21313C;
}

<span
    class="emotion-0 emotion-1"
  >
     → 
  </span>
  .emotion-0:last-of-type {
  color: #21313C;
}

.emotion-0:hover,
.emotion-0:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:hover:not(:last-of-type),
.emotion-0:focus:not(:last-of-type) {
  color: #21313C;
}

<a
    class="emotion-0 emotion-1"
    href="/view-analyze/"
  >
    View & Analyze Data
  </a>
</DocumentFragment>
`;
